File: plugin/src/main/java/com/google/tsunami/plugin/TcsClient.java
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.tsunami.plugin;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.tsunami.common.net.UrlUtils.removeTrailingSlashes;

import com.google.common.flogger.GoogleLogger;
import com.google.common.net.HostAndPort;
import com.google.common.net.InetAddresses;
import com.google.common.net.InternetDomainName;
import com.google.protobuf.util.JsonFormat;
import com.google.tsunami.callbackserver.common.CbidGenerator;
import com.google.tsunami.callbackserver.common.CbidProcessor;
import com.google.tsunami.callbackserver.common.Sha3CbidGenerator;
import com.google.tsunami.callbackserver.proto.PollingResult;
import com.google.tsunami.common.net.http.HttpClient;
import com.google.tsunami.common.net.http.HttpHeaders;
import com.google.tsunami.common.net.http.HttpRequest;
import com.google.tsunami.common.net.http.HttpResponse;
import java.io.IOException;
import java.util.Optional;

/** TcsClient for generating oob payload and retriving result */
public final class TcsClient {
  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
  private static final JsonFormat.Parser jsonParser = JsonFormat.parser();

  private final String callbackAddress;
  private final int callbackPort;
  private final String pollingBaseUrl;

  private final CbidGenerator cbidGenerator;
  private final HttpClient httpClient;

  public TcsClient(
      String callbackAddress, int callbackPort, String pollingBaseUrl, HttpClient httpClient) {
    this.callbackAddress = checkNotNull(callbackAddress);
    this.callbackPort = callbackPort;
    this.pollingBaseUrl = checkNotNull(pollingBaseUrl);
    this.cbidGenerator = new Sha3CbidGenerator();
    this.httpClient = checkNotNull(httpClient);
  }

  /**
   * Checks whether the callback server is configured. Detectors should use this to determine if
   * they can use the callback server for detecting vulnerabilities
   *
   * @return whether the callback server is enabled
   */
  public boolean isCallbackServerEnabled() {
    // only return false when all config fields are empty so that improper config (e.g., missing
    // certain fields) can be exposed. Note that {@link TcsClient} already checks that all the
    // class variables are not null.
    return !this.callbackAddress.isEmpty()
        && isValidPortNumber(this.callbackPort)
        && !this.pollingBaseUrl.isEmpty();
  }

  public String getCallbackUri(String secretString) {
    String cbid = cbidGenerator.generate(secretString);

    if (!isValidPortNumber(callbackPort)) {
      throw new AssertionError("Invalid callbackPort number specified");
    }

    HostAndPort hostAndPort =
        callbackPort == 80
            ? HostAndPort.fromHost(callbackAddress)
            : HostAndPort.fromParts(callbackAddress, callbackPort);

    // check if the specified address is raw IP or domain
    if (InetAddresses.isInetAddress(callbackAddress)) {
      return CbidProcessor.addCbidToUrl(cbid, hostAndPort);
    } else if (InternetDomainName.isValid(callbackAddress)) {
      return CbidProcessor.addCbidToSubdomain(cbid, hostAndPort);
    }
    // Should never reach here
    throw new AssertionError("Unrecognized address format, should be Ip address or valid domain");
  }

  public boolean hasOobLog(String secretString) {
    // making a blocking call to get result
    Optional<PollingResult> result = sendPollingRequest(secretString);

    if (result.isPresent()) {
      // In the future we may refactor hasOobLog() to return finer grained info about what kind
      // of oob is logged
      return result.get().getHasDnsInteraction() || result.get().getHasHttpInteraction();
    } else {
      // we may choose to retry sendPollingRequest() if oob interactions do arrive late.
      return false;
    }
  }

  private Optional<PollingResult> sendPollingRequest(String secretString) {
    HttpRequest request =
        HttpRequest.get(
                String.format("%s/?secret=%s", removeTrailingSlashes(pollingBaseUrl), secretString))
            .setHeaders(HttpHeaders.builder().addHeader("Cache-Control", "no-cache").build())
            .build();
    try {
      HttpResponse response = httpClient.send(request);
      if (response.status().isSuccess()) {
        PollingResult.Builder result = PollingResult.newBuilder();
        jsonParser.merge(response.bodyString().get(), result);
        return Optional.of(result.build());
      } else {
        logger.atInfo().log("OOB server returned %s", response.status().code());
      }
    } catch (IOException e) {
      logger.atWarning().withCause(e).log("Polling request failed");
    }
    return Optional.empty();
  }

  private static boolean isValidPortNumber(int port) {
    return port > 0 && port < 65536;
  }
}

File: plugin_server/py/plugin/tcs_client.py
"""Tsunami Callback Server client."""
import hashlib
from typing import Optional
from absl import logging
from net.proto2.python.public import json_format
from tsunami.plugin_server.py.common.data import network_endpoint_utils
from tsunami.plugin_server.py.common.net.http.http_client import HttpClient
from tsunami.plugin_server.py.common.net.http.http_headers import HttpHeaders
from tsunami.plugin_server.py.common.net.http.http_request import HttpRequest
from tsunami.proto import network_pb2
from tsunami_callbackserver.proto import polling_pb2


class TcsClient:
  """Client use for communicating with Tsunami Callback Server."""

  def __init__(
      self,
      callback_address: str,
      callback_port: int,
      polling_base_url: str,
      http_client: HttpClient,
  ):
    """Initialize Tsunami Callback Server client.

    Args:
      callback_address: IP address or the hostname of the callback server.
      callback_port: The port that the callback server is running on.
      polling_base_url: Base url of the callback server.
      http_client: Client used to send HTTP requests and process responses.
    """
    self.callback_endpoint = self._create_callback_address(
        callback_address, callback_port
    )
    self.polling_base_url = self._remove_trailing_slashes(polling_base_url)
    self.http_client = http_client

  def is_callback_server_enabled(self) -> bool:
    """Check if callback server is enabled.

    Returns:
      Whether callback server is reachable.
    """
    return bool(
        self.callback_endpoint.ip_address.address
        or self.callback_endpoint.hostname.name
    ) and bool(self.polling_base_url)

  def get_callback_uri(self, secret_string: str) -> str:
    """Assemble the URI to reach the callback server.

    Args:
      secret_string: Callback unique ID that is bonded to the scan target. Used
        to include in the request URI.

    Returns:
      The server URI.

      Examples with provided hostname:
        04041e8898e739ca33.google.com
        04041e8898e739ca33.google.com:8080

      Examples with provided IP address:
        http://127.0.0.1:8080/04041e8898e739ca33
        http://[2001:db8:3333:4444:5555:6666:7777:8888]/04041e8898e739ca33
    """
    # Generate hash from 8 bytes secret string using SHA-3 hashing
    cbid = hashlib.sha3_224(secret_string.encode('utf-8')).hexdigest()
    uri = network_endpoint_utils.to_uri_authority(self.callback_endpoint)
    # return uri with provided hostname
    if network_endpoint_utils.has_hostname(self.callback_endpoint):
      return '%s.%s' % (cbid, uri)
    # return uri with provided ip address
    return 'http://%s/%s' % (uri, cbid)

  def has_oob_log(self, secret_string: str) -> bool:
    """Check if callback server has received OOB log.

    Args:
      secret_string: Callback unique ID that is bonded to the scan target. Used
        to include in the request URI.

    Returns:
      If callback server has out-of-bounds log.
    """
    result = self._send_polling_request(secret_string)
    if result:
      return result.has_dns_interaction or result.has_http_interaction
    return False

  def _send_polling_request(
      self, secret_string: str
  ) -> Optional[polling_pb2.PollingResult]:
    """Send HTTP requests to the callback server.

    Args:
      secret_string: Callback unique ID that is bonded to the scan target. Used
        to include in the request URI.

    Returns:
      The polling results of whether or not the scan target has DNS or HTTP
      interaction. Returns None if polling request failed.
    """
    request = self._build_polling_request(secret_string)
    try:
      response = self.http_client.send(request)
      if response.status.is_success():
        return json_format.Parse(
            response.body_string(), polling_pb2.PollingResult()
        )
      else:
        logging.info('OOB server returned %s.', response.status.code)
    except (json_format.ParseError, ValueError):
      logging.exception('Polling request failed.')
    return None

  def _build_polling_request(self, secret_string: str) -> HttpRequest:
    url = '%s/?secret=%s' % (self.polling_base_url, secret_string)
    return (
        HttpRequest.get(url)
        .set_headers(
            HttpHeaders.builder()
            .add_header('Cache-Control', 'no-cache')
            .build()
        )
        .build()
    )

  def _create_callback_address(
      self, address: str, port: int
  ) -> network_pb2.NetworkEndpoint:
    try:
      return (
          network_endpoint_utils.for_ip(address)
          if port == 80
          else network_endpoint_utils.for_ip_and_port(address, port)
      )
    except ValueError:
      return (
          network_endpoint_utils.for_hostname(address)
          if port == 80
          else network_endpoint_utils.for_hostname_and_port(address, port)
      )

  def _remove_trailing_slashes(self, url: str) -> str:
    return url.strip('/')
